test/ERC827Token.js

```diff
@@ -0,0 +1,295 @@
+    await token.transferFrom(
+      accounts[0], accounts[2], balance0 + 1,
+      { from: accounts[1] }
+    ).should.be.rejectedWith(EVMRevert);
+  });
+
+  describe('validating allowance updates to spender', function () {
+    let preApproved;
+
+    it('should start with zero', async function () {
+      preApproved = await token.allowance(accounts[0], accounts[1]);
+      assert.equal(preApproved, 0);
+    });
+
+    it('should increase by 50 then decrease by 10', async function () {
+      await token.increaseApproval(accounts[1], 50);
+      let postIncrease = await token.allowance(accounts[0], accounts[1]);
+      preApproved.plus(50).should.be.bignumber.equal(postIncrease);
+      await token.decreaseApproval(accounts[1], 10);
+      let postDecrease = await token.allowance(accounts[0], accounts[1]);
+      postIncrease.minus(10).should.be.bignumber.equal(postDecrease);
+    });
+  });
+
+  it('should increase by 50 then set to 0 when decreasing by more than 50', async function () {
+    await token.approve(accounts[1], 50);
+    await token.decreaseApproval(accounts[1], 60);
+    let postDecrease = await token.allowance(accounts[0], accounts[1]);
+    postDecrease.should.be.bignumber.equal(0);
+  });
+
+  it('should throw an error when trying to transfer to 0x0', async function () {
+    await token.transfer(0x0, 100).should.be.rejectedWith(EVMRevert);
+  });
+
+  it('should throw an error when trying to transferFrom to 0x0', async function () {
+    await token.approve(accounts[1], 100);
+    await token.transferFrom(accounts[0], 0x0, 100, { from: accounts[1] })
+      .should.be.rejectedWith(EVMRevert);
+  });
+
+  describe('Test ERC827 methods', function () {
+
+    it('should return correct balances after transfer (with data) and show the event on receiver contract', async function () {
+      const message = await Message.new();
+
+      const extraData = message.contract.showMessage.getData(
+        web3.toHex(123456), 666, 'Transfer Done'
+      );
+
+      // Use method #8 tranfer of the abi to encode the data tx
+      const transferData = ethjsABI.encodeMethod(token.abi[8],
+        [message.contract.address, 100, extraData]
+      );
+      const transaction = await token.sendTransaction(
+        { from: accounts[0], data: transferData }
+      );
+
+      assert.equal(2, transaction.receipt.logs.length);
+
+      new BigNumber(100).should.be.bignumber.equal(
+        await token.balanceOf(message.contract.address)
+      );
+    });
+
+    it('should return correct allowance after approve (with data) and show the event on receiver contract', async function () {
+      const message = await Message.new();
+
+      const extraData = message.contract.showMessage.getData(
+        web3.toHex(123456), 666, 'Transfer Done'
+      );
+
+      // Use method #3 approve of the abi to encode the data tx
+      const approveData = ethjsABI.encodeMethod(token.abi[3],
+        [message.contract.address, 100, extraData]
+      );
```

Comment:
I understand that the encoding must be done manually due to truffle's limitations for function overloading. That being said, I'd rather `find` the desired function from the ABI array rather than relying on the ordering, as it is quite brittle; if we add another function later, these tests may break (or worse, may inadvertently test other functions)

Reply:
You are right! im going to change it :)